### ISSUE TITLE
Investigate if pruning works correctly

### ISSUE THREAD

**Issue by muXxer** [iotaledger/iota#4582]
It seems that the metrics `last_pruned_effects_checkpoint` metric always is `0` on our test node and also on some validator nodes of partners. We need to check if the pruning logic gets correctly executed.
Is the function `prune_checkpoints` correctly called?

This is the config used:
```
authority-store-pruning-config:
  num-latest-epoch-dbs-to-retain: 3
  epoch-db-pruning-period-secs: 3600
  num-epochs-to-retain: 1
  max-checkpoints-in-batch: 5
  max-transactions-in-batch: 1000
```

**Comment by gokhan-simsek-iota** [comment]
Just started running with using the `authority-store-pruning-config` above:

```
nohup ./target/release/iota-node --config-path fullnode.yaml > logs.rs 2>&1

tail -f logs.rs
...
2024-12-23T14:32:29.981798Z  WARN typed_store::rocks: very slow batch write elapsed=1.158617829 db_name="checkpoint"
2024-12-23T14:32:29.981830Z  WARN typed_store::rocks: very slow batch write elapsed=1.159495468 db_name="epoch"
2024-12-23T14:32:29.981980Z ERROR iota_metrics::thread_stall_monitor: Thread stalled for 0.95815125s. Possible causes include CPU overload or too much blocking calls.
2024-12-23T14:32:29.982130Z  WARN typed_store::rocks: very slow batch write elapsed=1.159264895 db_name="epoch"
2024-12-23T14:32:29.982260Z  WARN typed_store::rocks: very slow batch write elapsed=1.159700339 db_name="epoch"
2024-12-23T14:32:30.112807Z  INFO iota_core::checkpoints::checkpoint_executor: Checkpoint execution took 1.495551615s
2024-12-23T14:32:30.113013Z  INFO iota_core::checkpoints::checkpoint_executor: Finished syncing and executing checkpoint 1055000
```

**Comment by gokhan-simsek-iota** [comment]
@muXxer, not sure if I understand this metrics address, shouldn't it be a remote server IP there?

My full config:

```
cat testnet/fullnode.yaml

# Update this value to the location you want Iota to store its database
db-path: "/root/db/testnet"

# For ipv4, update this to "/ipv4/X.X.X.X/tcp/8080/http"
network-address: "/dns/localhost/tcp/8080/http"
metrics-address: "0.0.0.0:9184"
# this address is also used for web socket connections
json-rpc-address: "0.0.0.0:9000"
enable-event-processing: true

genesis:
  # Update this to the location of where the genesis file is stored
  genesis-file-location: "/root/testnet/genesis.blob"

# Update this to the location of where the migration file is stored
migration-tx-data-path: "/root/testnet/migration.blob"

authority-store-pruning-config:
  num-latest-epoch-dbs-to-retain: 3
  epoch-db-pruning-period-secs: 3600
  num-epochs-to-retain: 1
  max-checkpoints-in-batch: 5
  max-transactions-in-batch: 1000

p2p-config:
  seed-peers:
    - address: /dns/access-0.r.testnet.iota.cafe/udp/8084
      peer-id: 46064108d0b689ed89d1f44153e532bb101ce8f8ca3a3d01ab991d4dea122cfc
    - address: /dns/access-1.r.testnet.iota.cafe/udp/8084
      peer-id: 8ffd25fa4e86c30c3f8da7092695e8a103462d7a213b815d77d6da7f0a2a52f5
```

**Comment by gokhan-simsek-iota** [comment]
Checking the latest logs, for the last 3 minutes this is the only log getting produced:
(log file is ~600MB in 16hrs)
```
2024-12-24T06:27:46.266146Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 5295673
2024-12-24T06:27:46.268028Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 5295674
...
2024-12-24T06:29:28.164275Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 5305610
2024-12-24T06:29:28.164980Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 5305611
2024-12-24T06:29:28.165745Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 5305612
```

**Comment by gokhan-simsek-iota** [comment]
After 18hrs:
```
du -hs ~/db/testnet/
42G     /root/db/testnet/
```

**Comment by gokhan-simsek-iota** [comment]
Some logs from 3-day test run of fullnode for testnet:

```
// 3-day test - start: 2 Jan 2025 - end: still running

>>>> 2.1.2025
2025-01-02T10:50:57.182921Z  INFO iota_core::connection_monitor: Spawning future crates/iota-core/src/connection_monitor.rs:ConnectionMonitor
2025-01-02T10:50:57.182969Z  INFO iota_node: Iota chain identifier: 2304aa97
2025-01-02T10:50:57.183073Z  INFO iota_node::admin: starting admin server filter=info address=127.0.0.1:1337
2025-01-02T10:50:57.183211Z  INFO iota_core::checkpoints::checkpoint_executor: Checkpoint preparation for execution took 127.511µs
2025-01-02T10:50:57.183356Z  INFO iota_core::authority::authority_per_epoch_store: transaction write conflict detected, sleeping delay=17ms retries=1
2025-01-02T10:50:57.184105Z  INFO iota_core::checkpoints::checkpoint_executor: Checkpoint execution took 871.798µs
2025-01-02T10:50:57.184508Z  INFO iota_core::checkpoints::checkpoint_executor: Finished syncing and executing checkpoint 0
2025-01-02T10:51:05.033915Z  INFO iota_core::checkpoints::checkpoint_executor: Checkpoint preparation for execution took 105.161µs
2025-01-02T10:51:05.349683Z  INFO iota_core::checkpoints::checkpoint_executor: Checkpoint execution took 315.755242ms
2025-01-02T10:51:05.349827Z  INFO iota_core::checkpoints::checkpoint_executor: Finished syncing and executing checkpoint 5000
2025-01-02T10:51:13.681269Z  INFO iota_core::checkpoints::checkpoint_executor: Checkpoint preparation for execution took 247.7
.....
2025-01-02T12:19:37.004615Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 48076
2025-01-02T12:19:42.862942Z  INFO iota_core::checkpoints::checkpoint_executor: Checkpoint preparation for execution took 317.342µs
2025-01-02T12:19:43.184742Z  INFO iota_core::checkpoints::checkpoint_executor: Checkpoint execution took 321.784528ms
2025-01-02T12:19:43.184939Z  INFO iota_core::checkpoints::checkpoint_executor: Finished syncing and executing checkpoint 3140000
2025-01-02T12:19:46.943588Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 48077
2025-01-02T12:19:46.943777Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 48078
.....
20G     /root/db/testnet/

2025-01-02T15:25:47.383341Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 749733
2025-01-02T15:25:47.383594Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 749734
2025-01-02T15:25:51.608407Z  INFO iota_core::checkpoints::checkpoint_executor: Checkpoint preparation for execution took 1.675665ms
2025-01-02T15:25:52.024530Z  INFO iota_core::checkpoints::checkpoint_executor: Checkpoint execution took 416.105433ms
2025-01-02T15:25:52.024694Z  INFO iota_core::checkpoints::checkpoint_executor: Finished syncing and executing checkpoint 8590000
....
49G     /root/db/testnet/

>>>> 3.1.2025
2025-01-03T07:12:07.163795Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 8068718
2025-01-03T07:12:07.164037Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 8068719
2025-01-03T07:12:07.164114Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 8068720
....
62G     /root/db/testnet/

>>>> 4.1.2025
2025-01-04T07:24:47.017969Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 13893698
2025-01-04T07:24:47.018135Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 13893699
2025-01-04T07:24:50.690345Z  INFO iota_core::checkpoints::checkpoint_executor: Checkpoint preparation for execution took 392.234µs
2025-01-04T07:24:50.693246Z  INFO iota_core::checkpoints::checkpoint_executor: Checkpoint execution took 2.888425ms
2025-01-04T07:24:50.693451Z  INFO iota_core::checkpoints::checkpoint_executor: Finished syncing and executing checkpoint 17595000
2025-01-04T07:24:56.951760Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 13893700
2025-01-04T07:24:56.952044Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 13893701

....
60G     /root/db/testnet/


>>>> 5.1.2025
2025-01-05T07:03:16.980470Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 16232430
2025-01-05T07:03:20.144423Z  INFO iota_core::checkpoints::checkpoint_executor: Checkpoint preparation for execution took 459.494µs
2025-01-05T07:03:20.147726Z  INFO iota_core::checkpoints::checkpoint_executor: Checkpoint execution took 3.290379ms
2025-01-05T07:03:20.147935Z  INFO iota_core::checkpoints::checkpoint_executor: Finished syncing and executing checkpoint 17980000
2025-01-05T07:03:26.943503Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 16232431
2025-01-05T07:03:26.943637Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 16232432
...
61G     /root/db/testnet/
1.9G    /testnet/logs.rs

>>>> 6.1.2025
2025-01-06T07:55:06.970304Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 17387912
2025-01-06T07:55:06.970546Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 17387913
2025-01-06T07:55:06.970631Z  INFO iota_core::authority::authority_store_pruner: scheduling pruning for checkpoint 17387914
...
62G     /root/db/testnet/
2.1G    /testnet/logs.rs
```

**Comment by gokhan-simsek-iota** [comment]
Next actions:
- Find out how to see `last_pruned_effects_checkpoint` metric 
- Find out if the function prune_checkpoints correctly called
- I see that iota-tool cmd calls `prune_checkpoints`, I will check that also

**Comment by gokhan-simsek-iota** [comment]
Looks like `last_pruned_effects_checkpoint` is 0 for my full node
```
curl http://localhost:9184/metrics | grep last_pruned_effects_checkpoint
# TYPE last_pruned_effects_checkpoint gauge
last_pruned_effects_checkpoint 0
```

**Comment by gokhan-simsek-iota** [comment]
And `last_pruned_effects_checkpoint` is `0` for all validators, except for `iota-testnet-validator.nodes.guru:9184` which is `Last *: 17709364` 
Src: 
https://grafana.control.h.testnet.iota.cafe/d/d8c4099e-3560-4bc6-adb4-94380e0f602f/iota-rebased-overview?orgId=1&from=1736206449893&to=1736217249893&viewPanel=101

**Comment by gokhan-simsek-iota** [comment]
**Summary:** 
After adding `num_epochs_to_retain_for_checkpoints` into `fullnode,yaml`, `prune_checkpoints` was being called.

**Details:**
`prune_checkpoints` was NOT called in my tests, because looks like `num_epochs_to_retain_for_checkpoints` needs to be enabled for `prune_checkpoints` to be called. Here is the related code:

https://github.com/iotaledger/iota/blob/a15a205a5b753f62bc40312e2c3b41f054519dea/crates/iota-core/src/authority/authority_store_pruner.rs#L728-L731

To test this claim, I have added `num-epochs-to-retain-for-checkpoints` into `fullnode.yaml` and, restarted the node and I see in the logs that `prune_checkpoints` is being called and also the metric is not 0 anymore.

// fullnode.yaml
``` 
authority-store-pruning-config:
  ..
  num-epochs-to-retain-for-checkpoints: 2
  ..
```

// my local full node
```
curl http://localhost:9184/metrics | grep last_pruned_effects_checkpoint 
last_pruned_effects_checkpoint 282235

after 1 day 

curl http://localhost:9184/metrics | grep last_pruned_effects_checkpoint
last_pruned_effects_checkpoint 9038653
```

// iota-testnet-validator.nodes.guru
```
curl http://iota-testnet-validator.nodes.guru:9184/metrics | grep last_pruned_effects_checkpoint
last_pruned_effects_checkpoint 20147545
```

**Comment by muXxer** [comment]
@alexsporn so it turns out it was a configuration issue, pruning works if the config is correct.
